docs/doxygen/man/man3/GetBlock.3
.TH "GetBlock" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GetBlock \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGetBlock\fP (uint16_t src_bank, uint32_t src_user, uint32_t block, uint32_t time)"
.br
.ti -1c
.RI "\fBGetBlock\fP (const \fBGetBlock\fP &obj)=delete"
.br
.RI "\fIDisabled copy contructor\&. \fP"
.ti -1c
.RI "\fBGetBlock\fP & \fBoperator=\fP (const \fBGetBlock\fP &)=delete"
.br
.RI "\fIDisabled copy assignment operator\&. \fP"
.ti -1c
.RI "virtual \fB~GetBlock\fP ()"
.br
.RI "\fIFree responseBuffer\&. \fP"
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_NDS as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eReadingOnly as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet message id\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "virtual uint32_t \fBgetBlockId\fP ()"
.br
.RI "\fIGet block id\&. \fP"
.ti -1c
.RI "virtual \fBErrorCodes::Code\fP \fBprepareResponse\fP ()"
.br
.RI "\fIGet data from files and pack to data ready to send\&. \fP"
.ti -1c
.RI "void \fBprintSingleNode\fP (boost::property_tree::ptree &tree, int nodeId, \fBServersNode\fP &serverNode)"
.br
.RI "\fIPrints info for single node\&. \fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBGetBlockData\fP \fBm_data\fP"
.br
.ti -1c
.RI "\fBHelper::ServersHeader\fP \fBm_responseHeader\fP"
.br
.ti -1c
.RI "std::vector< \fBHelper::ServersNode\fP > \fBm_responseNodes\fP"
.br
.ti -1c
.RI "\fBHelper::Hlog\fP \fBm_hlog\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 11 of file getblock\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool GetBlock::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 72 of file getblock\&.cpp\&.
.SS "bool GetBlock::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 99 of file getblock\&.cpp\&.
.SS "void GetBlock::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 68 of file getblock\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/getblock.cpp
#include "getblock.h"

#include <netinet/in.h>
#include <arpa/inet.h>

#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/json.h"
#include "helper/ascii.h"
#include "helper/hlog.h"
#include "helper/blocks.h"

GetBlock::GetBlock()
    : m_data{}, m_hlog{} {
}

GetBlock::GetBlock(uint16_t src_bank, uint32_t src_user, uint32_t block, uint32_t time)
    : m_data(src_bank, src_user, block, time), m_hlog{} {
    if (!block) {
        m_data.info.block = time-(time%BLOCKSEC)-BLOCKSEC;
    }
}

GetBlock::~GetBlock() {

}

unsigned char* GetBlock::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* GetBlock::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_responseHeader);
}

void GetBlock::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void GetBlock::setResponse(char* response) {
    m_responseHeader = *reinterpret_cast<decltype(m_responseHeader)*>(response);
}

int GetBlock::getDataSize() {
    return sizeof(m_data.info);
}

int GetBlock::getResponseSize() {
    return sizeof(m_responseHeader);
}

unsigned char* GetBlock::getSignature() {
    return m_data.sign;
}

int GetBlock::getSignatureSize() {
    return sizeof(m_data.sign);
}

int GetBlock::getType() {
    return TXSTYPE_NDS;
}

CommandType GetBlock::getCommandType() {
    return CommandType::eReadingOnly;
}

void GetBlock::sign(const uint8_t* /*hash*/, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign(getData(), getDataSize(), sk, pk, getSignature());
}

bool GetBlock::checkSignature(const uint8_t* /*hash*/, const uint8_t* pk) {
    return (ed25519_sign_open(getData(), getDataSize(), pk, getSignature()) == 0);
}

void GetBlock::saveResponse(settings& /*sts*/) {
}

uint32_t GetBlock::getUserId() {
    return m_data.info.src_user;
}

uint32_t GetBlock::getBankId() {
    return m_data.info.src_node;
}

uint32_t GetBlock::getTime() {
    return m_data.info.ttime;
}

int64_t GetBlock::getFee() {
    return 0;
}

int64_t GetBlock::getDeduct() {
    return 0;
}

bool GetBlock::send(INetworkClient& netClient) {
    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("GetBlock sending error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("GetBlock reading error\n");
        return false;
    }

    if (m_responseError) {
        return true;
    }

    if(!netClient.readData(getResponse(), getResponseSize())) {
        ELOG("GetBlock ERROR reading response header info\n");
        return false;
    }

    for (unsigned int i=0; i<m_responseHeader.nodesCount; ++i) {
        Helper::ServersNode node;
        if (!netClient.readData((char*)&node, sizeof(node))) {
            ELOG("GetBlock ERROR reading nodes\n");
            return false;
        }
        m_responseNodes.push_back(node);
    }

    // read Hlog
    uint32_t hlogSize;
    if (!netClient.readData((int32_t*)&hlogSize, sizeof(uint32_t))) {
        ELOG("GetBlock Error reading Hlog size\n");
        return false;
    }

    if (hlogSize > 0) {
        char *hlogBuffer = new char[hlogSize];
        if (!netClient.readData(hlogBuffer, hlogSize)) {
            ELOG("GetBlock Error reading Hlog\n");
        }
        m_hlog = Helper::Hlog(hlogBuffer, hlogSize);
        delete[] hlogBuffer;
    }

    return true;
}

uint32_t GetBlock::getBlockId() {
    return m_data.info.block;
}

uint32_t GetBlock::getUserMessageId() {
    return 0;
}

ErrorCodes::Code GetBlock::prepareResponse() {
    uint32_t path = m_data.info.block;
    char filename[64];
    Helper::FileName::getName(filename, path, "servers.srv");
    Helper::Servers servers(filename);
    servers.load();

    if (servers.getNodesCount() > 0) {
        m_responseHeader = servers.getHeader();
        m_responseNodes = servers.getNodes();
    } else {
        return ErrorCodes::Code::eGetBlockInfoUnavailable;
    }
    return ErrorCodes::Code::eNone;
}

std::string GetBlock::toString(bool /*pretty*/) {
    return "";
}

void GetBlock::printSingleNode(boost::property_tree::ptree& tree, int nodeId, ServersNode& serverNode) {
    boost::property_tree::ptree node;

    char nodehex[5];
    nodehex[4]='\0';
    sprintf(nodehex,"%04X", nodeId);
    node.put("id", nodehex);

    serverNode.toJson(node);

    tree.push_back(std::make_pair("",node));
}

void GetBlock::toJson(boost::property_tree::ptree& ptree) {
    if (!m_responseError) {
        boost::property_tree::ptree blocktree;
        char blockhex[9];
        blockhex[8]='\0';

        sprintf(blockhex,"%08X", m_data.info.block);
        ptree.put("block_time_hex", blockhex);
        blocktree.put("id", blockhex);
        ptree.put("block_time", m_data.info.block);
        sprintf(blockhex,"%08X", m_data.info.block-BLOCKSEC);
        ptree.put("block_prev",blockhex);
        sprintf(blockhex,"%08X",m_data.info.block+BLOCKSEC);
        ptree.put("block_next",blockhex);

        m_responseHeader.toJson(blocktree);
        boost::property_tree::ptree nodes;
        for (unsigned int i=0; i<m_responseNodes.size(); ++i) {
            printSingleNode(nodes, i, m_responseNodes[i]);
        }
        blocktree.add_child("nodes", nodes);
        ptree.add_child("block", blocktree);

        if (m_hlog.total > 0) {
            m_hlog.printJson(ptree);
        }
    } else {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    }
}

void GetBlock::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.src_node);
    ptree.put(TAG::SRC_USER, m_data.info.src_user);
    ptree.put(TAG::BLOCK, m_data.info.block);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string GetBlock::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"get_block\",[\"block\":<block_time as hex>]}" << "\n";
    ss << "Example: " << "{\"run\":\"get_block\",\"block\":\"5B217880\"}" << "\n";
    return ss.str();
}
